src/Function/Metric/Definitions.agda

```diff
@@ -0,0 +1,60 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Definitions of properties over distance functions
+------------------------------------------------------------------------
+
+-- The contents of this module should be accessed via `Function.Metric`.
+
+{-# OPTIONS --without-K --safe #-}
+
+module Function.Metric.Definitions where
+
+open import Algebra.Core using (Op₂)
+open import Data.Product using (∃)
+open import Function.Metric.Core using (DistanceFunction)
+open import Level using (Level)
+open import Relation.Binary.Core using (Rel; _Preserves₂_⟶_⟶_)
+open import Relation.Nullary using (¬_)
+
+private
+  variable
+    a i ℓ ℓ₁ ℓ₂ : Level
+    A : Set a
+    I : Set i
+
+-----------------------------------------------------------------------
+-- Properties
+
+Congruent : Rel A ℓ₁ → Rel I ℓ₂ → DistanceFunction A I → Set _
+Congruent _≈ₐ_ _≈ᵢ_ d = d Preserves₂ _≈ₐ_ ⟶ _≈ₐ_ ⟶ _≈ᵢ_
+
+Indiscernable : Rel A ℓ₁ → Rel I ℓ₂ → DistanceFunction A I → I → Set _
+Indiscernable _≈ₐ_ _≈ᵢ_ d 0# = ∀ {x y} → d x y ≈ᵢ 0# → x ≈ₐ y
+
+PositiveDefinite : Rel A ℓ₁ → Rel I ℓ₂ → DistanceFunction A I → I → Set _
+PositiveDefinite _≈ₐ_ _≈ᵢ_ d 0# = ∀ {x y} → x ≈ₐ y → d x y ≈ᵢ 0#
+
+Symmetric : Rel I ℓ → DistanceFunction A I → Set _
+Symmetric _≈_ d = ∀ x y → d x y ≈ d y x
+
+TriangleInequality : Rel I ℓ → Op₂ I → DistanceFunction A I → _
+TriangleInequality _≤_ _∙_ d = ∀ x y z → d x z ≤ (d x y ∙ d y z)
+
+Bounded : Rel I ℓ → DistanceFunction A I → Set _
+Bounded _≤_ d = ∃ λ n → ∀ x y → d x y ≤ n
+
+TranslationInvariant : Rel I ℓ₂ → Op₂ A → DistanceFunction A I → Set _
+TranslationInvariant _≈_ _∙_ d = ∀ {x y a} → d (x ∙ a) (y ∙ a) ≈ d x y
+
+Contracting : Rel I ℓ → (A → A) → DistanceFunction A I → Set _
+Contracting _≤_ f d = ∀ x y → d (f x) (f y) ≤ d x y
```

Comment:
I somehow feel that `Contracting` on just a `Rel` rather than on a linear order feels wrong. The definition above is more like `Homomorphism` than `Contracting`.  I probably could have made the same comment about `Bounded`.

Reply:
I completely understand where you're coming from, however I think this is fairly standard practice in the library e.g. look at `Maximum` in `Relation.Binary.Definitions` in which we don't require the relation to be a partial order, or `Antisymmetric` in which we don't require the first relation to be a setoid.

The main reason for this is to avoid passing extraneous things into definitions. None of the proofs for a total order are used in the definition and neither is the equality relation we'd be forced to provide. This means that a) the definition can be used in environments where those assumptions are not yet available (even though they will be later on) and b) it provides better modularity of code. For example imagine we wanted a new structure `IsContractingMetric` which extended `IsMetric`. If we made the `contracting` definition require a total order, we'd have to first build that total order between the field `isMetric : IsMetric ...` and `contracting : Contracting ...` which would be inconvenient and ugly.

Reply:
I think you're right - so I think this means we ought to document that at the top level of the library somewhere.  Something along the lines of "Type abbreviations (such as `Maximum` and `Contracting`) are meant to capture the basic idea of certain definitions -- but are not meant to express the full concept on their own. The complete concepts are captured in various bundles where the abbreviation is correctly used in context. In other words, the bundles are meant to be the place where mathematical abstractions are adequately captured, while type abbreviations are meant to help modular reuse of the concepts that make up those abstractions."  Or something like that (my words above are probably a little too woolly).